compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -570,19 +570,9 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 self.dcx().emit_err(NeverPatternWithGuard { span: g.span });
             }
 
-            // We add a fake `loop {}` arm body so that it typecks to `!`.
-            // FIXME(never_patterns): Desugar into a call to `unreachable_unchecked`.
```

Comment:
Your PR description doesn't explain the challenges of just lowering a call to, for example, the `unreachable` intrinsic. I feel like that's far less invasive than adding a new HIR expr kind.